packages/number-field/src/vaadin-number-field.js

```diff
@@ -0,0 +1,435 @@
+        }
+
+        /* TODO: provide alternatives for Safari */
+        :host([dir='rtl']) [part='input-field'] ::slotted(input)::placeholder {
+          direction: rtl;
+        }
+
+        :host([dir='rtl']:not([has-controls])) [part='input-field'] ::slotted(input)::placeholder {
+          text-align: left;
+        }
+      </style>
+
+      <div part="container">
+        <div part="label" on-click="focus">
+          <slot name="label"></slot>
+        </div>
+
+        <vaadin-input-container
+          part="input-field"
+          readonly="[[readonly]]"
+          disabled="[[disabled]]"
+          invalid="[[invalid]]"
+          theme$="[[theme]]"
+        >
+          <div
+            disabled$="[[!_allowed(-1, value, min, max, step)]]"
+            part="decrease-button"
+            on-click="_decreaseValue"
+            on-touchend="_decreaseButtonTouchend"
+            hidden$="[[!hasControls]]"
+            slot="prefix"
+          ></div>
+          <slot name="prefix" slot="prefix"></slot>
+          <slot name="input"></slot>
+          <slot name="suffix" slot="suffix"></slot>
+          <div id="clearButton" part="clear-button" slot="suffix"></div>
+          <div
+            disabled$="[[!_allowed(1, value, min, max, step)]]"
+            part="increase-button"
+            on-click="_increaseValue"
+            on-touchend="_increaseButtonTouchend"
+            hidden$="[[!hasControls]]"
+            slot="suffix"
+          ></div>
+        </vaadin-input-container>
+
+        <div part="helper-text">
+          <slot name="helper"></slot>
+        </div>
+
+        <div part="error-message">
+          <slot name="error-message"></slot>
+        </div>
+      </div>
+    `;
+  }
+
+  static get properties() {
+    return {
+      /**
+       * Set to true to display value increase/decrease controls.
+       * @attr {boolean} has-controls
+       */
+      hasControls: {
+        type: Boolean,
+        value: false,
+        reflectToAttribute: true
+      },
+
+      /**
+       * The minimum value of the field.
+       */
+      min: {
+        type: Number,
+        observer: '_minChanged'
+      },
+
+      /**
+       * The maximum value of the field.
+       */
+      max: {
+        type: Number,
+        observer: '_maxChanged'
+      },
+
+      /**
+       * Specifies the allowed number intervals of the field.
+       * @type {number}
+       */
+      step: {
+        type: Number,
+        value: 1,
+        observer: '_stepChanged'
+      }
+    };
+  }
+
+  constructor() {
+    super();
+    this._setType('number');
+    this._boundOnChange = this.__onInputChange.bind(this);
+  }
+
+  /**
+   * Used by `ClearButtonMixin` as a reference to the clear button element.
+   * @protected
+   */
+  get clearElement() {
+    return this.$.clearButton;
+  }
+
+  /** @protected */
+  connectedCallback() {
+    super.connectedCallback();
+
+    if (this._inputNode) {
+      this._inputNode.min = this.min;
```

Comment:
Shouldn't it be initialized by the corresponding observer (`_minChanged`)?

Reply:
The problem is that `_minChanged` runs for the first time before `connectedCallback` in the following order:

1. `constructor`
2. `attributeChangedCallback` 
3. Observers are triggered
4. `connectedCallback`

Therefore in case of passing `min` and `max` using attributes, we need to forward them manually.

Reply:
Got it. Thanks!